docs/proposals/20220125-ipam-integration.md

```diff
@@ -0,0 +1,340 @@
+ memoryMiB: 8192
+ diskGiB: 45
+ network:
+   devices:
+   - dhcp4: false
+     fromPool: # reference to the pool
+       group: ipam.cluster.x-k8s.io/v1alpha1
+       kind: IPPool
+       name: testpool
+status:
+ network:
+   devices:
+     - claim:
+         name: example-1-1
+```
+
+### User Stories
+
+#### Story 1
+
+As a cluster operator I want to use static IP addresses with my Machines to avoid problems with DHCP lease duration during rolling cluster upgrades.
+
+#### Story 2
+
+As a cluster operator I want to integrate machines provisioned by CAPI with an external IPAM system and allocate IP addresses from it.
+
+#### Story 3
+
+As an infrastructure provider I want a single interface to integrate with multiple different IPAM providers.
+
+#### Story 4
+
+As an IPAM provider I want a single interface to integrate with all infrastructure providers that can benefit from IPAM integration.
+
+### Implementation Details/Notes/Constraints
+
+#### New API Types
+
+The following API Types should be added to cluster-api. In the first iteration they will be added to the experimental API and will therefore live in the `cluster.x-k8s.io` group.
+
+##### IPAddressClaim
+
+The IPAddressClaim is used to request ip addresses from a pool. It gets reconciled by IPAM providers, which can filter based on the `Spec.Pool` reference.
+
+After an IPAddressClaim is created, the `Spec.Pool` reference, and therefore essentially the entire `Spec` is immutable to simplify reconciliation of the resource.
+
+```go
+// IPAddressClaimSpec describes the desired state of an IPAddressClaim
+type IPAddressClaimSpec struct {
+   // Pool is a reference to the pool from which an IP address should be allocated.
+   Pool LocalObjectReference `json:"pool,omitempty"`
+}
+
+// IPAddressClaimStatus contains the status of an IPAddressClaim
+type IPAddressClaimStatus struct {
+   // Address is a reference to the address that was allocated for this claim.
+   Address LocalObjectReference `json:"address,omitempty"`
+
+  // Conditions provide details about the status of the claim.
+	// +optional
+	Conditions clusterv1.Conditions `json:"conditions,omitempty"`
+}
+ 
+// IPAddressClaim can be used to allocate IPAddresses from an IP Pool.
+type IPAddressClaim struct {
+   metav1.TypeMeta   `json:",inline"`
+   metav1.ObjectMeta `json:"metadata,omitempty"`
+ 
+   Spec   IPAddressClaimSpec   `json:"spec,omitempty"`
+   Status IPAddressClaimStatus `json:"status,omitempty"`
+}
+```
+
+##### IPAddress
+
+IPAddress resources are created to fulfil an IPAddressClaim. They are created by the IPAM provider that reconciles the claim.
+
+IPAddresses are immutable.
+
+```go
+// IPAddressSpec describes an IPAddress
+type IPAddressSpec struct {
+   // Claim is a reference to the claim this IPAddress was created for.
+   Claim LocalObjectReference `json:"claim,omitempty"`
+ 
+   // Pool is a reference to the pool that this IPAddress was created from.
+   Pool LocalObjectReference `json:"pool,omitempty"`
```

Comment:
do we envision use cases where a pool does not necessarily need to be presented as a CRD but something that's in handled but the provider transparently?

Reply:
For eventual consistency reasons I wouldn't validate the reference is valid. So in theory it could also point to nothing if that is actually a use case. But as some configuration is probably always required, a single pool shared between all clusters is probably the better option.